### PR TITLE
Add sol-netctl API to get extra service properties

### PR DESCRIPTION
**Pull request by guchaojie**
This PR is for to improve getting extra service properties for user
All the properties provided by Connman D-Bus Hierarchy

Signed-off-by: guchaojie <chao.jie.gu@intel.com>

### CODE REVIEW

**Comment by guchaojie** [comment]
@cabelitos @barbieri  Please have a review , thanks!

**Comment by guchaojie** [comment]
This PR also includes new APIs for getting extra service properties, please give some suggestion for these API definition @cabelitos @barbieri

File: data/scripts/libsoletta.sym [hunk 1]
```diff
@@ -605,6 +605,17 @@ global:
         sol_netctl_get_radios_offline;
         sol_netctl_get_services;
         sol_netctl_get_state;
+        sol_netctl_service_get_error;
```
Comment (cabelitos) [inline data/scripts/libsoletta.sym]: You should add a LIBSOLETTA_2 and add those new functions here.

Comment (cabelitos) [inline data/scripts/libsoletta.sym]: Check http://ftp.gnu.org/old-gnu/Manuals/ld-2.9.1/html_node/ld_25.html

Comment (cabelitos) [inline data/scripts/libsoletta.sym]: A real example: http://git.kernel.org/cgit/utils/kernel/kmod/kmod.git/tree/libkmod/libkmod.sym

File: src/lib/comms/include/sol-netctl.h [hunk 2]
```diff
@@ -153,6 +153,57 @@ enum sol_netctl_service_state {
...
+};
+
+/**
+ * @brief Structure to represent a network proxy.
+ *
+ * This struct contains the necessary information to present a
+ * network proxy for user.
+ */
+struct sol_proxy {
+    enum sol_proxy_method method;
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: I wonder if we need to an API version in here. Can this change in the future?

File: src/lib/comms/include/sol-netctl.h [hunk 3]
```diff
@@ -153,6 +153,57 @@ enum sol_netctl_service_state {
...
+};
+
+/**
+ * @brief Structure to represent a network provider.
+ *
+ * This struct contains the necessary information to present a
+ * network provider for user.
+ */
+struct sol_provider {
+    char *host;
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: Ditto

File: src/lib/comms/include/sol-netctl.h [hunk 4]
```diff
@@ -153,6 +153,57 @@ enum sol_netctl_service_state {
...
+};
+
+/**
+ * @brief Structure to represent a network ethernet.
+ *
+ * This struct contains the necessary information to present a
+ * network ethernet for user.
+ */
+struct sol_ethernet {
+    char *method;
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: Ditto

File: src/lib/comms/include/sol-netctl.h [hunk 5]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+ *
+ * This function gets the service favorite for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service favorite is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
+ */
+bool sol_netctl_service_get_favorite(
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: It's quite difficult to understand what you trying to achieve here. Do you want to get the favorite service or this function tells if the service itself is the favorite? If this function is used to check if the service in question is a favorite service or not I think we should rename to  sol_netlctl_service_is_favorite()

File: src/lib/comms/include/sol-netctl.h [hunk 6]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
 /**
+ * @brief Gets the service favorite
+ *
+ * This function gets the service favorite for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service favorite is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: Pay attention that will never return 0-100 or -errno. You're returning a bool value (true or false)

File: src/lib/comms/include/sol-netctl.h [hunk 7]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
     const struct sol_netctl_service *service);
 
 /**
+ * @brief Gets the service favorite
+ *
+ * This function gets the service favorite for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service favorite is
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: This sentence is quite diffucult to understand. I would change to something like "The service struct which will be used to check if it's the favorite" or something similar

File: src/lib/comms/include/sol-netctl.h [hunk 8]
```diff
@@ -268,6 +319,21 @@ enum sol_netctl_service_state sol_netctl_service_get_state(
...
 
 /**
+ * @brief Gets the service error
+ *
+ * This function gets the error for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the name is desired
+ *
+ * @return The error. When error occur during connection or disconnection
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: SHould you warn the user that if something goes wrong when trying to get the error (lol), this will return NULL. (Is this possible??)

File: src/lib/comms/include/sol-netctl.h [hunk 9]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+ *
+ * This function gets the service immutable for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service immutable is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
+ */
+bool sol_netctl_service_get_immutable(
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: All the same comments that I did about sol_netctl_service_get_favorite() applies to this function

File: src/lib/comms/include/sol-netctl.h [hunk 10]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+ *
+ * This function gets the service autoconnect for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service autoconnect is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
+ */
+bool sol_netctl_service_get_autoconnect(
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: All the same comments that I did about sol_netctl_service_get_favorite() applies to this function

File: src/lib/comms/include/sol-netctl.h [hunk 11]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+ *
+ * This function gets the service roaming for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service roaming is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
+ */
+bool sol_netctl_service_get_roaming(
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: All the same comments that I did about sol_netctl_service_get_favorite() applies to this function

File: src/lib/comms/include/sol-netctl.h [hunk 12]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+/**
+ * @brief Gets the service nameservers
+ *
+ * This function gets the service nameservers for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service nameservers is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: Be careful with copy & pasta. This function does not return an int!

File: src/lib/comms/include/sol-netctl.h [hunk 13]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+/**
+ * @brief Gets the service timeservers
+ *
+ * This function gets the service timeservers for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service timeservers is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: this function does not return an int!

File: src/lib/comms/include/sol-netctl.h [hunk 14]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+/**
+ * @brief Gets the service domains
+ *
+ * This function gets the service domains for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service domains is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: Ditto

File: src/lib/comms/include/sol-netctl.h [hunk 15]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+/**
+ * @brief Gets the service proxy
+ *
+ * This function gets the service proxy for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service proxy is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: Ditto

File: src/lib/comms/include/sol-netctl.h [hunk 16]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+ *
+ * This function gets the service provider for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service provider is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
+ */
+const struct sol_provider *
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: Ditto

File: src/lib/comms/include/sol-netctl.h [hunk 17]
```diff
@@ -309,6 +375,146 @@ int32_t sol_netctl_service_get_strength(
...
+ *
+ * This function gets the service ethernet for the netctl service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the service ethernet is
+ * desired
+ *
+ * @return 0-100 on success, -errno on failure.
+ */
+const struct sol_ethernet *
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: Ditto

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 18]
```diff
@@ -484,6 +964,118 @@ get_services_properties(sd_bus_message *m, const char *path)
...
+            const char *nameserver = NULL;
+
+            sol_ptr_vector_clear(&service->nameservers);
+
+            r = sd_bus_message_enter_container(m, 'a', "s");
+            SOL_INT_CHECK(r, < 0, r);
+
+            while (sd_bus_message_read_basic(m, 's', &nameserver) > 0) {
+                r = sol_ptr_vector_append(&service->nameservers, nameserver);
+                SOL_INT_CHECK(r, < 0, r);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: You should exit the container and you will leave the nameservers. I think you should iterate over the list the clear all the elements and them clear the list itself.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 19]
```diff
@@ -484,6 +964,118 @@ get_services_properties(sd_bus_message *m, const char *path)
...
+            const char *timeserver = NULL;
+
+            sol_ptr_vector_clear(&service->timeservers);
+
+            r = sd_bus_message_enter_container(m, 'a', "s");
+            SOL_INT_CHECK(r, < 0, r);
+
+            while (sd_bus_message_read_basic(m, 's', &timeserver) > 0) {
+                r = sol_ptr_vector_append(&service->timeservers, timeserver);
+                SOL_INT_CHECK(r, < 0, r);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: You should exit the container and you will leave the nameservers. I think you should iterate over the list the clear all the elements and them clear the list itself.

Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: You should exit the container and you will leave the nameservers. I think you should iterate over the list the clear all the elements and them clear the list itself.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 20]
```diff
@@ -484,6 +964,118 @@ get_services_properties(sd_bus_message *m, const char *path)
...
+            const char *domain = NULL;
+
+            sol_ptr_vector_clear(&service->domains);
+
+            r = sd_bus_message_enter_container(m, 'a', "s");
+            SOL_INT_CHECK(r, < 0, r);
+
+            while (sd_bus_message_read_basic(m, 's', &domain) > 0) {
+                r = sol_ptr_vector_append(&service->domains, domain);
+                SOL_INT_CHECK(r, < 0, r);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: You should exit the container and you will leave the nameservers. I think you should iterate over the list the clear all the elements and them clear the list itself.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 21]
```diff
@@ -484,6 +964,118 @@ get_services_properties(sd_bus_message *m, const char *path)
...
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else if (streq(str, "Roaming")) {
+            bool roaming;
+
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "b");
+            SOL_INT_CHECK(r, < 0, r);
+
+            r = sd_bus_message_read_basic(m, 'b', &roaming);
+            SOL_INT_CHECK(r, < 0, r);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: You should exit the container if something goes wrong, at least try to...

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 22]
```diff
@@ -484,6 +964,118 @@ get_services_properties(sd_bus_message *m, const char *path)
...
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else if (streq(str, "AutoConnect")) {
+            bool autoconnect;
+
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "b");
+            SOL_INT_CHECK(r, < 0, r);
+
+            r = sd_bus_message_read_basic(m, 'b', &autoconnect);
+            SOL_INT_CHECK(r, < 0, r);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: You should exit the container if something goes wrong, at least try to...

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 23]
```diff
@@ -484,6 +964,118 @@ get_services_properties(sd_bus_message *m, const char *path)
...
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else if (streq(str, "Immutable")) {
+            bool immutable;
+
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "b");
+            SOL_INT_CHECK(r, < 0, r);
+
+            r = sd_bus_message_read_basic(m, 'b', &immutable);
+            SOL_INT_CHECK(r, < 0, r);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: You should exit the container if something goes wrong, at least try to...

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 24]
```diff
@@ -484,6 +964,118 @@ get_services_properties(sd_bus_message *m, const char *path)
...
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else if (streq(str, "Favorite")) {
+            bool favorite;
+
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "b");
+            SOL_INT_CHECK(r, < 0, r);
+
+            r = sd_bus_message_read_basic(m, 'b', &favorite);
+            SOL_INT_CHECK(r, < 0, r);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: You should exit the container if something goes wrong, at least try to...

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 25]
```diff
@@ -484,6 +964,118 @@ get_services_properties(sd_bus_message *m, const char *path)
...
+            const char *security = NULL;
+
+            sol_ptr_vector_clear(&service->security);
+
+            r = sd_bus_message_enter_container(m, 'a', "s");
+            SOL_INT_CHECK(r, < 0, r);
+
+            while (sd_bus_message_read_basic(m, 's', &security) > 0) {
+                r = sol_ptr_vector_append(&service->security, security);
+                SOL_INT_CHECK(r, < 0, r);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: You should exit the container and you will leave the nameservers. I think you should iterate over the list the clear all the elements and them clear the list itself.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 26]
```diff
@@ -334,6 +490,316 @@ get_service_ip(sd_bus_message *m,
...
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    } while (1);
+
+end:
+    if (r == 0)
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: Why?

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 27]
```diff
@@ -334,6 +490,316 @@ get_service_ip(sd_bus_message *m,
...
+    return r;
+}
+
+static int
+get_ethernet_info(sd_bus_message *m, struct sol_ethernet *ethernet)
+{
+    char *str;
+    int r;
+
+    SOL_NULL_CHECK(ethernet, -EINVAL);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: This check is not necessary

File: src/lib/comms/sol-netctl-impl-connman.c (hunk 26 above)
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: I would change all the SOL_INT_CHECKs to jump to the end label.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 28]
```diff
@@ -334,6 +490,316 @@ get_service_ip(sd_bus_message *m,
...
+    return r;
+}
+
+static int
+get_provider_info(sd_bus_message *m, struct sol_provider *provider)
+{
+    char *str;
+    int r;
+
+    SOL_NULL_CHECK(provider, -EINVAL);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: Not necessary

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 29]
```diff
@@ -334,6 +490,316 @@ get_service_ip(sd_bus_message *m,
...
+            SOL_DBG("Ignored service provider property: %s", str);
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    } while (1);
+
+end:
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: Again, I would change too SOL_INT_CHECK to jump to this label.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 30]
```diff
@@ -334,6 +490,316 @@ get_service_ip(sd_bus_message *m,
...
+            SOL_DBG("Ignored service proxy property: %s", str);
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    } while (1);
+
+end:
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: Again, I would change too SOL_INT_CHECK to jump to this label.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 31]
```diff
@@ -334,6 +490,316 @@ get_service_ip(sd_bus_message *m,
     return r;
 }
 
+static int
+get_proxy_info(sd_bus_message *m, struct sol_proxy *proxy)
+{
+    char *str;
+    int r;
+
+    SOL_NULL_CHECK(proxy, -EINVAL);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: Not necessary

File: src/lib/comms/include/sol-netctl.h [hunk 32]
```diff
@@ -153,6 +153,57 @@ enum sol_netctl_service_state {
...
+    SOL_PROXY_METHOD_MANUAL,
+};
+
+/**
+ * @brief Structure to represent a network proxy.
+ *
+ * This struct contains the necessary information to present a
+ * network proxy for user.
+ */
+struct sol_proxy {
```
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: I guess we're using typedefs for structs now. You should add them as well

**Comment by cabelitos** [comment]
@guchaojie Sorry for the delay. I did a initial review. Most problems are repeat, it should be easy to fix. :]

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 33]
```diff
@@ -500,7 +1092,31 @@ get_services_properties(sd_bus_message *m, const char *path)
...
+
+            get_provider_info(m, &service->provider);
+
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else if (streq(str, "Ethernet")) {
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "a{sv}");
+            SOL_INT_CHECK(r, < 0, r);
+
+            get_ethernet_info(m, &service->ethernet);
```
Comment (cabelitos) [inline src/lib/comms/sol-netctl-impl-connman.c]: I realized that you're not checking the return value of this function (and other get_*). You should check it here, so we're able to report the error... (Do the same for get_service_ip in another patch, please)

File: src/lib/comms/include/sol-netctl.h (hunk 2 above)
Comment (guchaojie) [inline src/lib/comms/include/sol-netctl.h]: Of course we can change . What kind of change do you refer to ?
BTW, which API version define here?

File: src/lib/comms/include/sol-netctl.h (hunk 5 above)
Comment (guchaojie) [inline src/lib/comms/include/sol-netctl.h]: Thanks for your suggestion.  API function name should be more clear here.

File: src/lib/comms/include/sol-netctl.h (hunk 2 above)
Comment (cabelitos) [inline src/lib/comms/include/sol-netctl.h]: Something like this:https://github.com/solettaproject/soletta/blob/master/src/lib/comms/include/sol-lwm2m.h#L357
